### PR TITLE
Return an exception with a meaningful message for all errors that might occur during downloading forms

### PR DESCRIPTION
**Pull request by grzesiek2010**
Work towards #4489 

#### What has been done to verify that this works as intended?
I added some automated test and tested the changes manually hardcoding some exceptions.

#### Why is this the best possible solution? Were any other approaches considered?
The problem was that the errors that might occur during downloading forms were turned into generic `FormDownloadException` so even if `FormSourceException`[ was caught with its message it was turned into `FormDownloadException` with no message](https://github.com/getodk/collect/pull/4824/files#diff-edbeccead389ffa8df15f518180a4aefdfa0f7fc236abd975b86dc3920d41c3cL116).
I changed that and also added some new errors like: https://github.com/getodk/collect/pull/4824/files#diff-4237726580b2c992e7730f2d4c0e0d9c5899d97c7c263bb8f905906f11ba6189R15 so that every exception has its own message.

I decided also  to remove `FormDownloadException`  and add those new exceptions in `FormSourceException` becasue to me it makes more sense so have one hierarchy of exceptions related to downloading forms.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
It doesn't require testing, let's test it together with UI changes.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: forms/src/main/java/org/odk/collect/forms/FormSourceException.kt
Comment (seadowg) [inline forms/src/main/java/org/odk/collect/forms/FormSourceException.kt]: Looking at this now, it does feel a little strange to me that `FormSourceException` includes the download specific errors. This way these exceptions are declared in `forms` but never come out of any components defined in there. Do you think we could move back to a `FormDownloadException` hierarchy for the new errors and just have `FormDownloader` throw both? I still think having one mapper makes sense, so it could have a `getMessage` for both super types.

File: collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormDownloaderTest.java
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormDownloaderTest.java]: We could make this more specific now and just catch a `FormWithNoHashException`. Same goes for the other `catch (FormSourceException e)` tests below.

File: forms/src/main/java/org/odk/collect/forms/FormSourceException.kt
Comment (seadowg) [reply forms/src/main/java/org/odk/collect/forms/FormSourceException.kt]: Naming wise, I don't think these need the `...Exception` postfix as they are all inner classes, i.e. they'll be addressed as `FormSourceException.X`.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormDownloaderTest.java
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/formmanagement/ServerFormDownloaderTest.java]: Fixed.

File: forms/src/main/java/org/odk/collect/forms/FormSourceException.kt
Comment (grzesiek2010) [reply forms/src/main/java/org/odk/collect/forms/FormSourceException.kt]: > it does feel a little strange to me that FormSourceException includes the download specific errors

hmm now as I think about it maybe you are right. I'm not strongly convinced but let's do that.

> Naming wise, I don't think these need the ...Exception postfix as they are all inner classes, i.e. they'll be addressed as FormSourceException.X.

Fixed.

**Review by seadowg — APPROVED** [review]
This will definitely add a lot more detail to the download errors. Let's make sure to review the error text when we have the new UI for displaying these.

Going to wait on merging this until #4803 is merged so we can include that PR in a beta without this.

**Pull request merged**
